### ISSUE TITLE
Update using r.literal() in nested structure fails (?)

### ISSUE THREAD

**Issue by tatsujin1** [rethinkdb/rethinkdb#5977]
I have a database migration script that (amongst other things) converts a "flat" value into structured value. In my case I'm changing an id referencing a document in another table to storing the other document "inline".

Anyway, to specifics.

Given the following document:
```
{
   'id': 'test', 
   'data': 'flat_value' 
}
```
And then doing an update, replacing `data` with a structured value that contains `r.literal()` (and in my case a lot of other things).
Performing the write works fine. Or at least appears to:
```
>> table.get('test').update({
   'data': {
       'nested': r.literal()
    }
}).run()
{
   'replaced': 1,
   'skipped': 0,
   'unchanged': 0,
   'inserted': 0,
   'deleted': 0,
   'errors': 0
}
```

No error. However, attempting to retrieve the document fails:

```
>> table.get('test').run()
ReqlDriverError: Unknown pseudo-type LITERAL
```

This happens only when _replacing_ an existing value with an object that contains r.literal().
Given the original document and adding a new field works fine:
```
>> table.get('test').update({
   'new_data': {
      'nested': r.literal()
   }
}).run()
>> table.get('test').run()
{
   'id': 'test',
   'data': 'flat_value',
   'new_data': { }
}
```

edit: shortened queries
edit2: fixed output of last query

**Comment by danielmewes** [comment]
Thanks for reporting this @tatsujin1 .

That's definitely a bug. The `literal` should get removed before the document gets stored...

**Comment by danielmewes** [comment]
Another sequence of queries to test when fixing this (reported by @tve in #5978 ):
```js
r.db("crs").table("resources").delete()

r.db("crs").table("resources").get("raZsCqcfQGTLj7WhpqcquA").replace(function(var_29) {
  return { id: "raZsCqcfQGTLj7WhpqcquA", bag: var_29("bag").default({}).merge(null)}
}, {durability: "soft", return_changes:true, non_atomic:true})

r.db("crs").table("resources").get("raZsCqcfQGTLj7WhpqcquA").update({
  bag: {red: r.literal({colors:{top:"blu", bot:"grn"}})}
}, {durability: "soft", return_changes:true, non_atomic:true})
```

**Comment by mlucy** [comment]
This is probably caused by `merge` only recursing until it sees the first new field, and then adding everything under that field.  The easiest way to fix this would be to make `merge` recurse down all the way through the thing it's merging, but that would make `merge` much less efficient in some cases in order to address a fairly rare edge case.

Two other options:
1. Do the merge the way we do it now, and change the code so that a `LITERAL` outside the righthand side of a merge is just ignored and treated the same as a missing field.
2. Tag objects to indicate whether any of their children contain a `LITERAL`.
  - This will stop us from having to traverse the whole object at merge-time, but will slightly increase overhead for most other operations on objects.

I think `1` would perform the best (and would also solve any similar related bugs in the future), but it's also hardest to implement.

**Comment by danielmewes** [comment]
I think 1. would be fairly easy to implement if we modify it slightly. If we treat a `LITERAL` value outside of a `merge` as a `null` value, rather than a missing field, we could modify the `datum_t` class to do this transparently and without much performance penalty. The `getField` functions would simply return a `null` datum instead of the actual field value if they detect the field value to be a `LITERAL`.

Treating them as missing fields is harder to fit into the existing API. Currently you iterate over fields in a `datum_t` by getting the total number of fields first, and then requesting each one of them by its index if I remember correctly. If we want to treat `LITERAL`s as missing fields, we would have to scan all fields before returning the size of the datum, which would be expensive. Furthermore, randomly accessing a field by index would no longer work efficiently.

Treating `LITERAL` fields as `null` is not completely consistent with the usual meaning of `LITERAL`, so I'm not sure if this would actually be a good idea.

A third option would be giving up the short-circuiting during a `merge`, and filtering out `LITERAL`s during a merge explicitly. This would make merges slightly less efficient when nested objects are involved and only present on one side of the merge, but I don't think it would be that bad.

**Comment by tve** [comment]
I'm probably not understanding what you mean by "a LITERAL outside the righthand side of a merge is just ignored" but it sounds like a really bad idea to me. I'm understanding that this implies that my use case simply wouldn't work? What I'm trying to do: I have a nested object that may or may not have a specific field and I'd like to "upsert" that field in its entirety. Hence something like `{foo: {bar: r.literal() } }`, i.e. here bar may or may not exist. You're proposing that in this case bar would not get created if it didn't exist? I would have to wrap this into something like
```
r.branch(r.and(row.hasField("foo"), row.("foo").hasField("bar")),
{foo: { bar: <value> }},
{foo: { bar: r.literal(<value>) }})
```
(syntax most likely slightly off)

**Comment by danielmewes** [comment]
@tve Are you actually using `r.literal()` here without any arguments? That would imply that the field should get deleted. Or are you actually passing something into `r.literal` to replace the field if it exists, and insert it otherwise?

Our discussion above was mostly assuming an empty `r.literal`. Non-empty `r.literal` would need to be handled like the values that are inside of them indeed.

@mlucy Actually I'm now thinking that "hiding" `LITERAL`s when reading from a `datum` might be a bad idea. It might complicate comparing datums, and seems bug-prone. I think the best option might be just traversing both sides completely during a `merge` and stripping out the datum when doing that. I think we already have a function that does that. We just need to call it in a few additional cases.

**Comment by danielmewes** [comment]
"stripping out the datum" => "stripping out the literals"

**Comment by tve** [comment]
@danielmewes my use-case has values inside the r.literal. You pasted my exact queries above https://github.com/rethinkdb/rethinkdb/issues/5977#issuecomment-234027824

**Comment by mlucy** [comment]
When I say "A literal outside the righthand side of a merge is just ignored", I mean that `{a: foo}` and `{a: r.literal(foo)}` are identical, and also that `{a: 1, b: r.literal()}` and `{a: 1}` are identical.

@danielmewes -- I think treating `r.literal()` fields as `NULL` would be really bad.  Unless I'm misunderstanding the proposal, `merge` as it's currently implemented would cause `r({}).merge({foo: {bar: r.literal()}})` to return `{foo: {bar: r.literal()}}` which would be `{foo: {bar: NULL}}`, while `r({foo: {}}).merge({foo: {bar: r.literal()}})` would return `{foo: {})`.

---

I don't think it would be a huge hassle to change the object interface so that you use an iterator to access the fields instead, and you can pass in an option indicating whether you want the iterator to see empty `r.literal()`s or not (which only `merge` and maybe like one other thing would).

After thinking about it, though, I think the cleanest option would actually be to tag objects with whether or not they contain a nested literal.  I think we currently have (or should have) the rule that objects returned by executing ReQL never contain `r.literal()`, so we'd only have to tag literal objects provided by the query, which means it won't slow things down much (if at all) and won't be that much new code.

**Comment by danielmewes** [comment]
> After thinking about it, though, I think the cleanest option would actually be to tag objects with whether or not they contain a nested literal.

What do you think about the option of just traversing the whole object on a `merge` to strip out `literal`s? I think that's probably just one or two lines of code (since we already have the function that does it), and I think in practice the performance penalty isn't going to be that big. It will only affect the case where one side of the merge adds a nested object that doesn't exist on the other side (and maybe it will affect nested arrays always? I'm not sure what we're currently doing for those).

Adding a flag isn't as easy, because it will also change the serialization format for datums.

**Comment by mlucy** [comment]
> Adding a flag isn't as easy, because it will also change the serialization format for datums.

We could implement it as an additional datum type (we'd have `OBJECT` and `OBJECT_WITH_LITERAL` or something).  That should be backward-compatible with on-disk data.

> What do you think about the option of just traversing the whole object on a merge to strip out literals?

It's probably fine, actually; you're right that it's not that much more expensive in the common case if we implement it well (i.e. doing it during the merge traversal rather than as a preprocessing step that generates a new object).

**Comment by danielmewes** [comment]
A fix for this is in internal code review 3712 .

**Comment by danielmewes** [comment]
Fixed in next 652c8fc and v2.3.x 563a403.

**Comment by danielmewes** [comment]
@tatsujin1 The fix for this was included in RethinkDB 2.3.5, which we released today.
